collector/report.go

```diff
@@ -0,0 +1,331 @@
+		platform, ok := annotations["platform"].(string)
+		if ok {
+			a.Platform = platform
+		}
+	}
+	if err := a.Validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Validate checks that we have a valid report submitted
+func (a *ActiveReport) Validate() error {
+	if testNameRegexp.MatchString(a.TestName) != true {
+		return errors.New("invalid \"test_name\" field")
+	}
+	if probeASNRegexp.MatchString(a.ProbeASN) != true {
+		return errors.New("invalid \"probe_asn\" field")
+	}
+	if probeCCRegexp.MatchString(a.ProbeCC) != true {
+		return errors.New("invalid \"probe_cc\" field")
+	}
+	if stringInSlice(a.Format, supportedFormats) != true {
+		return fmt.Errorf("invalid \"format\" field, %s", a.Format)
+	}
+	return nil
+}
+
+// IncomingFilename determines the filename of an incoming report
+func (a *ActiveReport) IncomingFilename() string {
+	if stringInSlice(a.Format, supportedFormats) != true {
+		// Defensive coding
+		panic("an unexpected value for format was found. Bailing...")
+	}
+	return fmt.Sprintf("%s.%s",
+		a.ReportID,
+		a.Format)
+}
+
+// SyncFilename creates a filename in the sync directory of the active report
+func (a *ActiveReport) SyncFilename() (string, error) {
+	// Defensive coding.
+	if err := a.Validate(); err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%s-%s-%s-%s-%s-probe-0.2.0.%s",
+		a.CreationTime.Format(TimestampFormat),
+		a.TestName,
+		a.ReportID,
+		a.ProbeASN,
+		a.ProbeCC,
+		a.Format), nil
+}
+
+// BackendExtra is serverside extra metadata
+type BackendExtra struct {
+	SubmissionTime time.Time `json:"submission_time" bson:"submission_time"`
+	MeasurementID  string    `json:"measurement_id" bson:"measurement_id"`
+	ReportID       string    `json:"report_id" bson:"report_id"`
+}
+
+// MeasurementEntry is the structure of measurements submitted by an OONI Probe client
+type MeasurementEntry struct {
+	// These values are added by the pipeline
+	// BucketDate     string `json:"bucket_date"`
+	// ReportFilename string `json:"report_filename"`
+
+	ID                   string       `json:"id"`
+	ReportID             string       `json:"report_id"`
+	TestName             string       `json:"test_name"`
+	TestVersion          string       `json:"test_version"`
+	MeasurementStartTime string       `json:"measurement_start_time,omitempty"`
+	TestStartTime        string       `json:"test_start_time"` // XXX these should actually be time
+	Annotations          interface{}  `json:"annotations"`
+	BackendExtra         BackendExtra `json:"backend_extra"`
+	BackendVersion       string       `json:"backend_version"`
+	DataFormatVersion    string       `json:"data_format_version"`
+	Input                string       `json:"input"`
+	InputHashes          []string     `json:"input_hashes"`
+	Options              []string     `json:"options"`
+	ProbeASN             string       `json:"probe_asn"`
+	ProbeCC              string       `json:"probe_cc"`
+	ProbeCity            string       `json:"probe_city"`
```

Comment:
Also, `ProbeCity` may be `null` (and MK always sets it to `null`).

Reply:
Good point. Probably a similar strategy to `input` needs to be applied to this as well.

Reply:
Right. I am wondering whether we care deeply about this field. Also, I think we can probably have a bump in the data version where we define this as strictly a string.